File: cdm/cpp/system/equipment/anesthesiamachine/actions/SEAnesthesiaMachineAction.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include "stdafx.h"
#include "system/equipment/anesthesiamachine/actions/SEAnesthesiaMachineAction.h"

SEAnesthesiaMachineAction::SEAnesthesiaMachineAction() : SEAction()
{

}

SEAnesthesiaMachineAction::~SEAnesthesiaMachineAction()
{
  Clear();
}

void SEAnesthesiaMachineAction::Clear()
{
  SEAction::Clear();
}

bool SEAnesthesiaMachineAction::IsValid() const
{
  return SEAction::IsValid();
}

void SEAnesthesiaMachineAction::Serialize(const cdm::AnesthesiaMachineActionData& src, SEAnesthesiaMachineAction& dst)
{
  dst.Clear();
}

void SEAnesthesiaMachineAction::Serialize(const SEAnesthesiaMachineAction& src, cdm::AnesthesiaMachineActionData& dst)
{

}

#include "system/equipment/anesthesiamachine/actions/SEAnesthesiaMachineConfiguration.h"
#include "system/equipment/anesthesiamachine/actions/SEOxygenWallPortPressureLoss.h"
#include "system/equipment/anesthesiamachine/actions/SEOxygenTankPressureLoss.h"
#include "system/equipment/anesthesiamachine/actions/SEExpiratoryValveLeak.h"
#include "system/equipment/anesthesiamachine/actions/SEExpiratoryValveObstruction.h"
#include "system/equipment/anesthesiamachine/actions/SEInspiratoryValveLeak.h"
#include "system/equipment/anesthesiamachine/actions/SEInspiratoryValveObstruction.h"
#include "system/equipment/anesthesiamachine/actions/SEMaskLeak.h"
#include "system/equipment/anesthesiamachine/actions/SESodaLimeFailure.h"
#include "system/equipment/anesthesiamachine/actions/SETubeCuffLeak.h"
#include "system/equipment/anesthesiamachine/actions/SEVaporizerFailure.h"
#include "system/equipment/anesthesiamachine/actions/SEVentilatorPressureLoss.h"
#include "system/equipment/anesthesiamachine/actions/SEYPieceDisconnect.h"
SEAnesthesiaMachineAction* SEAnesthesiaMachineAction::Load(const cdm::AnyAnesthesiaMachineActionData& any, SESubstanceManager& subMgr)
{
  switch (any.Action_case())
  {
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kConfiguration:
    {
      SEAnesthesiaMachineConfiguration* a = new SEAnesthesiaMachineConfiguration(subMgr);
      SEAnesthesiaMachineConfiguration::Load(any.configuration(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kOxygenTankPressureLoss:
    {
      SEOxygenTankPressureLoss* a = new SEOxygenTankPressureLoss();
      SEOxygenTankPressureLoss::Load(any.oxygentankpressureloss(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kOxygenWallPortPressureLoss:
    {
      SEOxygenWallPortPressureLoss* a = new SEOxygenWallPortPressureLoss();
      SEOxygenWallPortPressureLoss::Load(any.oxygenwallportpressureloss(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kExpiratoryValveLeak:
    {
      SEExpiratoryValveLeak* a = new SEExpiratoryValveLeak();
      SEExpiratoryValveLeak::Load(any.expiratoryvalveleak(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kExpiratoryValveObstruction:
    {
      SEExpiratoryValveObstruction* a = new SEExpiratoryValveObstruction();
      SEExpiratoryValveObstruction::Load(any.expiratoryvalveobstruction(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kInspiratoryValveLeak:
    {
      SEInspiratoryValveLeak* a = new SEInspiratoryValveLeak();
      SEInspiratoryValveLeak::Load(any.inspiratoryvalveleak(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kInspiratoryValveObstruction:
    {
      SEInspiratoryValveObstruction* a = new SEInspiratoryValveObstruction();
      SEInspiratoryValveObstruction::Load(any.inspiratoryvalveobstruction(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kMaskLeak:
    {
      SEMaskLeak* a = new SEMaskLeak();
      SEMaskLeak::Load(any.maskleak(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kSodaLimeFailure:
    {
      SESodaLimeFailure* a = new SESodaLimeFailure();
      SESodaLimeFailure::Load(any.sodalimefailure(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kTubeCuffLeak:
    {
      SETubeCuffLeak* a = new SETubeCuffLeak();
      SETubeCuffLeak::Load(any.tubecuffleak(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kVaporizerFailure:
    {
      SEVaporizerFailure* a = new SEVaporizerFailure();
      SEVaporizerFailure::Load(any.vaporizerfailure(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kVentilatorPressureLoss:
    {
      SEVentilatorPressureLoss* a = new SEVentilatorPressureLoss();
      SEVentilatorPressureLoss::Load(any.ventilatorpressureloss(), *a);
      return a;
    }
    case cdm::AnyAnesthesiaMachineActionData::ActionCase::kYPieceDisconnect:
    {
      SEYPieceDisconnect* a = new SEYPieceDisconnect();
      SEYPieceDisconnect::Load(any.ypiecedisconnect(), *a);
      return a;
    }
  }
  subMgr.Error("Unknown action type : " + any.Action_case());
  return nullptr;
}
cdm::AnyAnesthesiaMachineActionData* SEAnesthesiaMachineAction::Unload(const SEAnesthesiaMachineAction& action)
{
  cdm::AnyAnesthesiaMachineActionData* any = new cdm::AnyAnesthesiaMachineActionData();
  const SEAnesthesiaMachineConfiguration* amc = dynamic_cast<const SEAnesthesiaMachineConfiguration*>(&action);
  if (amc != nullptr)
  {
    any->set_allocated_configuration(SEAnesthesiaMachineConfiguration::Unload(*amc));
    return any;
  }

  const SEOxygenWallPortPressureLoss* owppl = dynamic_cast<const SEOxygenWallPortPressureLoss*>(&action);
  if (owppl != nullptr)
  {
    any->set_allocated_oxygenwallportpressureloss(SEOxygenWallPortPressureLoss::Unload(*owppl));
    return any;
  }

  const SEOxygenTankPressureLoss* otpl = dynamic_cast<const SEOxygenTankPressureLoss*>(&action);
  if (otpl != nullptr)
  {
    any->set_allocated_oxygentankpressureloss(SEOxygenTankPressureLoss::Unload(*otpl));
    return any;
  }

  const SEExpiratoryValveLeak* evl = dynamic_cast<const SEExpiratoryValveLeak*>(&action);
  if (evl != nullptr)
  {
    any->set_allocated_expiratoryvalveleak(SEExpiratoryValveLeak::Unload(*evl));
    return any;
  }

  const SEExpiratoryValveObstruction* evo = dynamic_cast<const SEExpiratoryValveObstruction*>(&action);
  if (evo != nullptr)
  {
    any->set_allocated_expiratoryvalveobstruction(SEExpiratoryValveObstruction::Unload(*evo));
    return any;
  }

  const SEInspiratoryValveLeak* ivl = dynamic_cast<const SEInspiratoryValveLeak*>(&action);
  if (ivl != nullptr)
  {
    any->set_allocated_inspiratoryvalveleak(SEInspiratoryValveLeak::Unload(*ivl));
    return any;
  }

  const SEInspiratoryValveObstruction* ivo = dynamic_cast<const SEInspiratoryValveObstruction*>(&action);
  if (ivo != nullptr)
  {
    any->set_allocated_inspiratoryvalveobstruction(SEInspiratoryValveObstruction::Unload(*ivo));
    return any;
  }

  const SEMaskLeak* ml = dynamic_cast<const SEMaskLeak*>(&action);
  if (ml != nullptr)
  {
    any->set_allocated_maskleak(SEMaskLeak::Unload(*ml));
    return any;
  }

  const SESodaLimeFailure* slf = dynamic_cast<const SESodaLimeFailure*>(&action);
  if (slf != nullptr)
  {
    any->set_allocated_sodalimefailure(SESodaLimeFailure::Unload(*slf));
    return any;
  }

  const SETubeCuffLeak* tcl = dynamic_cast<const SETubeCuffLeak*>(&action);
  if (tcl != nullptr)
  {
    any->set_allocated_tubecuffleak(SETubeCuffLeak::Unload(*tcl));
    return any;
  }

  const SEVaporizerFailure* vf = dynamic_cast<const SEVaporizerFailure*>(&action);
  if (vf != nullptr)
  {
    any->set_allocated_vaporizerfailure(SEVaporizerFailure::Unload(*vf));
    return any;
  }

  const SEVentilatorPressureLoss* vpl = dynamic_cast<const SEVentilatorPressureLoss*>(&action);
  if (vpl != nullptr)
  {
    any->set_allocated_ventilatorpressureloss(SEVentilatorPressureLoss::Unload(*vpl));
    return any;
  }

  const SEYPieceDisconnect* ypd = dynamic_cast<const SEYPieceDisconnect*>(&action);
  if (ypd != nullptr)
  {
    any->set_allocated_ypiecedisconnect(SEYPieceDisconnect::Unload(*ypd));
    return any;
  }
  delete any;
  return nullptr;
}

File: cdm/java/mil/tatrc/physiology/datamodel/system/equipment/anesthesia/actions/SEAnesthesiaMachineAction.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.system.equipment.anesthesia.actions;


import org.jfree.util.Log;

import com.kitware.physiology.cdm.AnesthesiaMachineActions.AnesthesiaMachineActionData;
import com.kitware.physiology.cdm.AnesthesiaMachineActions.AnyAnesthesiaMachineActionData;

import mil.tatrc.physiology.datamodel.actions.SEAction;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;
import mil.tatrc.physiology.datamodel.system.environment.actions.SEThermalApplication;

public abstract class SEAnesthesiaMachineAction extends SEAction
{
  public SEAnesthesiaMachineAction()
  {
    
  }
  
  public void copy(SEAnesthesiaMachineAction other)
  {
    if (this == other)
      return;
    super.copy(other);
  }
  
  public void reset()
  {
    super.reset();
  }
  
  public static void load(AnesthesiaMachineActionData src, SEAnesthesiaMachineAction dst) 
  {
    SEAction.load(src.getAction(), dst);
  }
  protected static void unload(SEAnesthesiaMachineAction src, AnesthesiaMachineActionData.Builder dst)
  {
    SEAction.unload(src, dst.getActionBuilder());
  }
  
  public static SEAnesthesiaMachineAction ANY2CDM(AnyAnesthesiaMachineActionData c, SESubstanceManager subMgr) 
  {
    switch(c.getActionCase())
    {
    case CONFIGURATION:
    {
      SEAnesthesiaMachineConfiguration dst = new SEAnesthesiaMachineConfiguration();
      SEAnesthesiaMachineConfiguration.load(c.getConfiguration(), dst, subMgr);
      return dst;
    }
    case EXPIRATORYVALVELEAK:
    {
      SEExpiratoryValveLeak dst = new SEExpiratoryValveLeak();
      SEExpiratoryValveLeak.load(c.getExpiratoryValveLeak(), dst);
      return dst;
    }
    case EXPIRATORYVALVEOBSTRUCTION:
    {
      SEExpiratoryValveObstruction dst = new SEExpiratoryValveObstruction();
      SEExpiratoryValveObstruction.load(c.getExpiratoryValveObstruction(), dst);
      return dst;
    }
    case INSPIRATORYVALVELEAK:
    {
      SEInspiratoryValveLeak dst = new SEInspiratoryValveLeak();
      SEInspiratoryValveLeak.load(c.getInspiratoryValveLeak(), dst);
      return dst;
    }
    case INSPIRATORYVALVEOBSTRUCTION:
    {
      SEInspiratoryValveObstruction dst = new SEInspiratoryValveObstruction();
      SEInspiratoryValveObstruction.load(c.getInspiratoryValveObstruction(), dst);
      return dst;
    }
    case MASKLEAK:
    {
      SEMaskLeak dst = new SEMaskLeak();
      SEMaskLeak.load(c.getMaskLeak(), dst);
      return dst;
    }
    case SODALIMEFAILURE:
    {
      SESodaLimeFailure dst = new SESodaLimeFailure();
      SESodaLimeFailure.load(c.getSodaLimeFailure(), dst);
      return dst;
    }
    case TUBECUFFLEAK:
    {
      SETubeCuffLeak dst = new SETubeCuffLeak();
      SETubeCuffLeak.load(c.getTubeCuffLeak(), dst);
      return dst;
    }
    case VAPORIZERFAILURE:
    {
      SEVaporizerFailure dst = new SEVaporizerFailure();
      SEVaporizerFailure.load(c.getVaporizerFailure(), dst);
      return dst;
    }
    case VENTILATORPRESSURELOSS:
    {
      SEVentilatorPressureLoss dst = new SEVentilatorPressureLoss();
      SEVentilatorPressureLoss.load(c.getVentilatorPressureLoss(), dst);
      return dst;
    }
    case YPIECEDISCONNECT:
    {
      SEYPieceDisconnect dst = new SEYPieceDisconnect();
      SEYPieceDisconnect.load(c.getYPieceDisconnect(), dst);
      return dst;
    }
    case OXYGENTANKPRESSURELOSS:
    {
      SEOxygenTankPressureLoss dst = new SEOxygenTankPressureLoss();
      SEOxygenTankPressureLoss.load(c.getOxygenTankPressureLoss(), dst);
      return dst;
    }
    case OXYGENWALLPORTPRESSURELOSS:
    {
      SEOxygenWallPortPressureLoss dst = new SEOxygenWallPortPressureLoss();
      SEOxygenWallPortPressureLoss.load(c.getOxygenWallPortPressureLoss(), dst);
      return dst;
    }
    case ACTION_NOT_SET:
      Log.warn("AnyAnesthesiaMachineActionData was empty...was that intended?");
      return null;
    }
    Log.error("Unsupported AnesthesiaMachine Action type "+c.getActionCase());
    return null;
  }
  public static AnyAnesthesiaMachineActionData CDM2ANY(SEAnesthesiaMachineAction c)
  {
    AnyAnesthesiaMachineActionData.Builder dst = AnyAnesthesiaMachineActionData.newBuilder();
    if(c instanceof SEAnesthesiaMachineConfiguration)
    {
      dst.setConfiguration(SEAnesthesiaMachineConfiguration.unload((SEAnesthesiaMachineConfiguration)c));
      return dst.build();
    }
    if(c instanceof SEExpiratoryValveLeak)
    {
      dst.setExpiratoryValveLeak(SEExpiratoryValveLeak.unload((SEExpiratoryValveLeak)c));
      return dst.build();
    }
    if(c instanceof SEExpiratoryValveObstruction)
    {
      dst.setExpiratoryValveObstruction(SEExpiratoryValveObstruction.unload((SEExpiratoryValveObstruction)c));
      return dst.build();
    }
    if(c instanceof SEInspiratoryValveLeak)
    {
      dst.setInspiratoryValveLeak(SEInspiratoryValveLeak.unload((SEInspiratoryValveLeak)c));
      return dst.build();
    }
    if(c instanceof SEInspiratoryValveObstruction)
    {
      dst.setInspiratoryValveObstruction(SEInspiratoryValveObstruction.unload((SEInspiratoryValveObstruction)c));
      return dst.build();
    }
    if(c instanceof SEMaskLeak)
    {
      dst.setMaskLeak(SEMaskLeak.unload((SEMaskLeak)c));
      return dst.build();
    }
    if(c instanceof SESodaLimeFailure)
    {
      dst.setSodaLimeFailure(SESodaLimeFailure.unload((SESodaLimeFailure)c));
      return dst.build();
    }
    if(c instanceof SETubeCuffLeak)
    {
      dst.setTubeCuffLeak(SETubeCuffLeak.unload((SETubeCuffLeak)c));
      return dst.build();
    }
    if(c instanceof SEVaporizerFailure)
    {
      dst.setVaporizerFailure(SEVaporizerFailure.unload((SEVaporizerFailure)c));
      return dst.build();
    }
    if(c instanceof SEVentilatorPressureLoss)
    {
      dst.setVentilatorPressureLoss(SEVentilatorPressureLoss.unload((SEVentilatorPressureLoss)c));
      return dst.build();
    }
    if(c instanceof SEYPieceDisconnect)
    {
      dst.setYPieceDisconnect(SEYPieceDisconnect.unload((SEYPieceDisconnect)c));
      return dst.build();
    }
    if(c instanceof SEOxygenTankPressureLoss)
    {
      dst.setOxygenTankPressureLoss(SEOxygenTankPressureLoss.unload((SEOxygenTankPressureLoss)c));
      return dst.build();
    }
    if(c instanceof SEOxygenWallPortPressureLoss)
    {
      dst.setOxygenWallPortPressureLoss(SEOxygenWallPortPressureLoss.unload((SEOxygenWallPortPressureLoss)c));
      return dst.build();
    }
    Log.error("Unsupported AnesthesiaMachine Action type "+c);
    return dst.build();
  }
  
  public abstract String toString();
}
